CORE-1212 Add Metadata Template Viewer and Listing

psarando opened the pull request:
This PR will import and update the Metadata Template Viewer dialog from the cyverse-de/ui repo, and adds a Metadata Template listing dialog (previously only in the GXT code).

This will also add the 3rd party endpoint handlers for the `Ontology Lookup Service` and `Unified Astronomy Thesaurus` services, used by the corresponding metadata template search fields.

Metadata Template listing dialog:
![Metadata Template listing dialog](https://user-images.githubusercontent.com/996408/106689909-07780680-658e-11eb-8ce4-437d2b230236.png)

Mobile view:
![Metadata Template listing dialog - mobile](https://user-images.githubusercontent.com/996408/106689979-2bd3e300-658e-11eb-8ad6-49e0d6359de2.png)

---

If a Template has a `description` defined, then the listing will include the `i` info icon-button next to the template download button. Select the `i` info button will show the template's description in a popover:
![Metadata Template description popover](https://user-images.githubusercontent.com/996408/106690143-7ce3d700-658e-11eb-9d4f-c9ed383947eb.png)

Mobile view:
![Metadata Template description popover - mobile](https://user-images.githubusercontent.com/996408/106690184-9422c480-658e-11eb-9bed-215b9c961c98.png)

---

The Metadata Template dialog looks essentially the same as in the DE, except cleaned up or standardized for Sonora: no more App bar in the dialog header (it's a `DEDialog` now) and the dialog's action buttons are now `Cancel` and `Save`:
![Metadata Template dialog](https://user-images.githubusercontent.com/996408/106690419-009dc380-658f-11eb-920f-aa4b192ba034.png)

Mobile view:
![Metadata Template dialog - mobile](https://user-images.githubusercontent.com/996408/106690514-2dea7180-658f-11eb-9475-ddb80ff1f78a.png)

The OLS and UAT search fields are still using the old Autocomplete field from `ui-lib`:
![Ontology Lookup Service search field](https://user-images.githubusercontent.com/996408/106692296-87ec3680-6591-11eb-8ece-5e52ef80882a.png)

They should be updated with the newer `@material-ui/lab/Autocomplete` field used by Sonora, but that will require refactoring the `FormSearchField` in `ui-lib`.

sri2k1us commented on src/components/metadata/templates/Listing.js:
```diff
@@ -0,0 +1,275 @@
+    );
+};
+
+export default MetadataTemplateListing;
```
I prefer this listing in a `List` view rather than in a table. Then we can eliminate an additional click for `description`(`i` button) and put description as `secondary` text prop and `download` action as `secondary` action. See examples here - https://material-ui.com/components/lists/
Regardless the buttons needs in a theme color rather than `black`.

psarando replied:
Good idea. I think the `List` looks better:
![Select Template dialog](https://user-images.githubusercontent.com/996408/107101476-b1e16b00-67d4-11eb-9f22-b835b346153e.png)

Mobile:
![Select Template dialog on mobile](https://user-images.githubusercontent.com/996408/107101513-d3425700-67d4-11eb-80ee-897e0c2bda56.png)

sri2k1us commented on public/static/locales/en/metadata.json:
```diff
@@ -10,6 +10,8 @@
     "attrTypeLabel": "Type",
     "avus": "AVUs",
     "back": "Back",
+    "confirmDiscardChangesDialogHeader": "Discard Changes?",
+    "confirmDiscardChangesDialogMsg": "You have unsaved changes. Do you wish to continue without saving?",
```
When I trying this PR locally, I ran into confusion here with the generic `Ok` and `Cancel` action button. We should change this to prompt the user for `saving` the changes rather than `discarding`. 
`"confirmSaveChangesDialogHeader": "Save Changes?",
"confirmSaveChangesDialogMsg": "You have unsaved changes. Do you wish to save now?"`
and the dialog options should be `Save`, `Discard` and `Cancel` rather than generic `OK` and `Cancel`.

sri2k1us replied:
I don't know if we are doing this consistently throughout Sonora, if not we should start doing this (i.e providing specific actions instead of generic `OK` and `Cancel`).

psarando replied:
I seem to remember having this discussion before where we actually reached the opposite conclusion. I think in the old DE we had 3 buttons like that for some dialogs, but we decided it was too much cognitive load on the user, and we decided on only 1 or 2 buttons for cancel/confirm or cancel/ok instead. I think this was also a recommendation from the Material spec. I will see if I can find a reference to that discussion again.

sri2k1us replied:
Ahh I do remember now something about having 3 buttons. But I still recommending having actions specific buttons instead of generic `Ok / Cancel`.

psarando replied:
I couldn't find where we had that discussion, but I did find on the Material spec they recommend a max of 2 buttons: https://material.io/components/dialogs#actions (under the `Number of actions` section).

I also agree the button text could be more specific 👍
That's recommended by the Material spec as well (https://material.io/components/dialogs#alert-dialog).

I think we should stick with the `Cancel` / `Discard` buttons, though, because this confirmation dialog only shows when the user is trying to close the template without saving, so we're confirming that choice. Otherwise they should `Cancel` the discard action and specifically select the `Save` button (which is an action that can add or modify many AVUs at a time).

sri2k1us replied:
Sounds good to me 👍🏽

psarando replied:
I also updated the content text, so the `Discard` button with dialog's message is hopefully clearer:
`You have unsaved changes. Do you wish to continue and discard these changes?`

sri2k1us reviewed (approved):
LGTM 🎉 . I had some comments on UX issues.

sri2k1us commented:
Ohh forgot to mention, we have been using `text` variant buttons for actions in dialog. I see all the dialogs in this PR use `contained` variant buttons. That needs to changed as well.

psarando commented on src/server/api/metadata.js:
```diff
@@ -7,15 +7,49 @@
 import { handler as terrainHandler } from "./terrain";
+import { metadataTemplateCSVhandler } from "./downloads";
 
 export default function notificationsRouter() {
```
Oops, looks like we all missed this in #238 😆

psarando commented:
I think I've addressed all PR comments, so I'll go ahead and merge this now, but additional reviews are still welcome 🙇 

Side note: the Metadata Template UX was one of the first (large) components we built with React and MUI, so I'm sure there is still room for improvement in a future PR. Perhaps lessons learned in the new App Editor could be applied to Metadata Templates in the future as well.

ash3rz reviewed (commented):
LGTM :shipit:
